Markdown Support & Improved formatting

ziyad-m97 opened the pull request:
## Overview

This PR introduces **full Markdown rendering support** in EULER's chat interface, along with improved UX for connector loading states and better backend response formatting.

---

## Markdown Support

### Android App Changes

#### New Components (`app/src/main/java/com/android/sample/ui/components/MarkdownText.kt`)

#### Chat Integration (`ChatMessage.kt`)
- AI messages now render with `MarkdownText` instead of plain `Text`
- Moodle overview content also uses Markdown rendering

### Backend Changes (`functions/src/index.ts`)

#### Updated `EPFL_SYSTEM_PROMPT`
Instructs EULER to use Markdown formatting appropriately 

#### Format by Query Type
- 📅 **Schedule**: Table-like format with bold days/times, room codes in backticks
- 🍴 **Food/Menu**: Restaurant name bold, bullet list of dishes with prices
- 📚 **EPFL Info (RAG)**: Structured with key facts, bullet summaries
- 📖 **Moodle**: Course name bold, file names in code, clear bullet list
- 💬 **ED Discussion**: Quote relevant answers, summarize with bullet points

## UX Improvements: Loading Indicators

### Moodle Loading
- **Problem**: Raw JSON was visible during streaming until parsing completed
- **Solution**: Detect incomplete Moodle JSON by content pattern and show friendly loading
- **Message**: 🟠 *"Parsing your Moodle courses…"* with animated orange dot
<img width="200" height="90" alt="image" src="https://github.com/user-attachments/assets/851e8fc6-9d1c-432f-bedb-537529408f0f" />

### ED Discussion Loading 
- Enhanced loading state with animated indicator
- **Message**: 🟣 *"Searching ED Discussion…"* with animated purple dot
<img width="203" height="151" alt="image" src="https://github.com/user-attachments/assets/f9c04986-64c9-42ac-a09c-d83a6559d91f" />


---

## Files Changed

### New Files
- `app/src/main/java/com/android/sample/ui/components/MarkdownText.kt`
- `app/src/test/java/com/android/sample/ui/components/MarkdownTextTest.kt`

### Modified Files
- `gradle/libs.versions.toml` - Added richtext library
- `app/build.gradle.kts` - Added richtext dependencies
- `app/src/main/java/com/android/sample/Chat/ChatMessage.kt` - Markdown + Moodle loading
- `app/src/main/java/com/android/sample/home/EdPostsSection.kt` - ED loading indicator
- `app/src/androidTest/java/com/android/sample/Chat/ChatMessageTest.kt` - Test fixes
- `functions/src/index.ts` - System prompt + source URL fix
- `functions/src/intentDetection.ts` - Moodle patterns
- `functions/src/moodleIntent.ts` - Ordinal support + course extraction


#### NB : I have used Gen AI to help me tackle these formatting and RegEX issues.

ziyad-m97 committed:
feat: add Markdown rendering support for chat messages

- Integrated richtext library for Markdown rendering in chat messages.
- Updated ChatMessage composable to utilize MarkdownText for rendering AI messages.
- Added MarkdownText component for consistent Markdown styling across the app.
- Included unit tests for MarkdownText functionality to ensure proper rendering of various Markdown elements.

ziyad-m97 committed:
Merge main into feat/rendering-markdown

Resolved conflict in ChatMessage.kt by keeping the refactored
structure (UserMessageRow, AiMessageColumn) from main while
integrating MarkdownText rendering for AI messages.

ziyad-m97 committed:
formatted files

treaks0 commented on app/src/main/java/com/android/sample/ui/components/MarkdownText.kt:
```diff
@@ -0,0 +1,339 @@
+ * markdown parsing isn't needed. Handles: bold, italic, inline code, links
+ */
+@Composable
+fun SimpleMarkdownText(
```
I like the idea of having a lighter SimpleMarkdownText for basic formatting.Since we now have both MarkdownText and SimpleMarkdownText, could you add a short documentation explaining when to use one vs the other

treaks0 commented on functions/src/intentDetection.ts:
Good here also. The added regexes make the Moodle file-fetch intent more flexible. Nice work.

treaks0 commented on functions/src/moodleIntent.ts:
Nice improvements here, the ordinal handling, expanded regex patterns, and clearer extraction logic make the Moodle intent detection much more robust.

ziyad-m97 committed:
fix(test): Update ChatMessageTest for MarkdownText rendering

The test was checking for literal "- Topic" text, but MarkdownText
renders the "- " prefix as a bullet point. Updated to check for
"Topic" substring instead.

ziyad-m97 committed:
docs(markdown): Improve MarkdownText documentation and remove unused params

- Add detailed comments explaining when to use MarkdownText vs SimpleMarkdownText
- Remove unused textColor and style parameters from MarkdownText
- Update ChatMessage.kt to match new MarkdownText signature

treaks0 reviewed (changes requested):
Overall, this is a really solid PR
You’ve cleaned up the prompts, switched our answers to consistent markdown formatting, and made Moodle intent detection much more robust (ordinals, course name patterns, etc.), plus added tests for the new markdown components.
I’ve left a few small comments/nits, but nothing blocking . once those are addressed and tests are green, I’m happy to approve

ziyad-m97 committed:
fix(chat): Show friendly loading message while parsing Moodle data

Instead of showing raw JSON during streaming, display a friendly
"Parsing your Moodle courses…" message with an animated indicator.
This improves UX by giving users clear feedback about what's happening.

ziyad-m97 committed:
fix(chat): Detect incomplete Moodle JSON regardless of streaming flag

The previous fix relied on isStreaming flag which may already be false
when tokens arrive. Now we detect Moodle JSON by content pattern:
- If text starts with { and contains "type":"moodle"
- And JSON parsing fails (incomplete)
- Show loading indicator instead of raw JSON

ziyad-m97 committed:
feat(ui): Add friendly loading indicator for ED Discussion search

Added animated "Searching ED Discussion…" message with purple dot
while fetching posts, matching the Moodle loading indicator style
for consistent UX across connectors.

ziyad-m97 committed:
format : formatted files.

ziyad-m97 committed:
fix(test): Update ChatMessageRobolectricTest for MarkdownText rendering

Same fix as ChatMessageTest - use substring matching for "Topic"
instead of assertTextEquals for literal "- Topic" since MarkdownText
renders the dash as a bullet point.

ziyad-m97 committed:
Merge branch 'main' into feat/rendering-markdown

ziyad-m97 committed:
format: formatted files

ziyad-m97 committed:
fix(home): Prevent overwriting local messages when no conversation is selected and incoming messages are empty

ziyad-m97 committed:
fix(home): fix race condition that failed test.

ziyad-m97 committed:
test: added comprehensive testing for markdown formatting.

ziyad-m97 committed:
feat: Extracted 2 helpers functions to eliminate code duplication.

ziyad-m97 committed:
enhancing: reduced cognitive complexity to tackle high sonar issues

ziyad-m97 committed:
format: formatted kotlin files

treaks0 reviewed (approved):
Great job ziyad , all the changes have been carried out.
Ready to merge.
